eventmesh-connectors/eventmesh-connector-http/src/main/java/org/apache/eventmesh/connector/http/source/protocol/impl/CloudEventProtocol.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.eventmesh.connector.http.source.protocol.impl;
+
+import org.apache.eventmesh.connector.http.common.BoundedConcurrentQueue;
+import org.apache.eventmesh.connector.http.source.config.SourceConnectorConfig;
+import org.apache.eventmesh.connector.http.source.protocol.Protocol;
+import org.apache.eventmesh.openconnect.offsetmgmt.api.data.ConnectRecord;
+import org.apache.eventmesh.openconnect.util.CloudEventUtil;
+
+import io.cloudevents.CloudEvent;
+import io.cloudevents.http.vertx.VertxMessageFactory;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.ext.web.Route;
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * CloudEvent Protocol.
+ */
+@Slf4j
+public class CloudEventProtocol implements Protocol {
+
+    // Protocol name
+    public static final String PROTOCOL_NAME = "CloudEvent";
+
+
+    /**
+     * Initialize the protocol.
+     *
+     * @param sourceConnectorConfig source connector config
+     */
+    @Override
+    public void initialize(SourceConnectorConfig sourceConnectorConfig) {
+
+    }
+
+
+    /**
+     * Handle the protocol message for CloudEvent.
+     *
+     * @param route     route
+     * @param boundedQueue queue info
+     */
+    @Override
+    public void setHandler(Route route, BoundedConcurrentQueue<Object> boundedQueue) {
+        route.method(HttpMethod.POST)
+            .handler(ctx -> VertxMessageFactory.createReader(ctx.request())
+                .map(reader -> {
+                    CloudEvent event = reader.toEvent();
+                    if (event.getSubject() == null) {
+                        throw new IllegalStateException("attribute 'subject' cannot be null");
+                    }
+                    if (event.getDataContentType() == null) {
+                        throw new IllegalStateException("attribute 'datacontenttype' cannot be null");
+                    }
+                    if (event.getData() == null) {
+                        throw new IllegalStateException("attribute 'data' cannot be null");
+                    }
+                    return event;
```

Comment:
Do you mean that the Webhook client should support CloudEvent and send a payload with `subject`, `datacontenttype` fields in it?

Reply:
This is not the case. As mentioned above, CloudEventProtocol is just a protocol compatible with older versions. As for whether CloudEventProtocol can be used as a protocol for webhook, I think there is no problem at all, because to make Source Connector support webhook, it only needs to have the corresponding webhook protocol implementation and may require a longer idleTimeout. In design, all protocols are mutually exclusive, that is, one Source Connector can only correspond to one Protocol, as shown in the figure below.

![Uploading image.png…]()